go/connection-refresh/btrefresh/bigtable_rotator.go

```diff
@@ -0,0 +1,93 @@
+package btrefresh
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/bigtable"
+)
+
+// Time to wait before killing the previous connection
+// during rotation. This may kill ongoing requests if they
+// take longer than this amount of time.
+const lameduckTime = 1 * time.Minute
+
+// RotatingTable is a bigtable.Table that automatically
+// reconnects to Cloud Bigtable at a given interval.
+type RotatingTable struct {
+	*bigtable.Table
+	client *bigtable.Client
+	swap   *time.Ticker
+	errors chan error
+}
+
+// BackgroundErrors is a channel that will propagate errors
+// that come from refreshing the connection. It can either be
+// checked with select before performing ops or watched
+// in a background thread.
+func (r RotatingTable) BackgroundErrors() <-chan error {
+	return r.errors
+}
+
+// Close will close the connections and stop rotating.
+func (r RotatingTable) Close() {
+	r.swap.Stop()
+	close(r.errors)
+	r.client.Close()
+}
+
+// BtDialer encapsulates making a connection to your bigtable.
+// This way you can use any means for dialing/ getting credentials.
+type BtDialer func() (*bigtable.Client, error)
+
+// NewRotatingTable makes a new Table reference that will refresh itself in the background at the given
+// interval.
+func NewRotatingTable(dialer BtDialer, table string, refresh time.Duration) (*RotatingTable, error) {
+	client, err := dialer()
+	errors := make(chan error, 1)
+	if err != nil {
+		return nil, err
+	}
+	tbl := client.Open(table)
+	warmTable(tbl)
+	ticker := time.NewTicker(refresh)
+	rt := &RotatingTable{tbl, client, ticker, errors}
+	go func() {
+		for _ = range ticker.C {
+			// Close the old client after waiting a bit
+			go func() {
+				oldC := rt.client
+				time.Sleep(lameduckTime)
+				oldC.Close()
+			}()
+			client, err := dialer()
+			if err != nil {
+				errors <- err
+				continue
+			}
+			tbl := client.Open(table)
+			warmTable(tbl)
+			rt.Table = tbl
+		}
+	}()
+	return rt, nil
+}
+
+// A new table reference will need to have some requests
+// go across it to establish each connection in the connection
+// pool.
+func warmTable(tbl *bigtable.Table) {
+	wg := sync.WaitGroup{}
+	// Run the warming requests across threads to saturate the
+	// connection pool.
+	for i := 0; i < 32; i++ {
```

Comment:
[Putting comments here from previous PR]
garye
Why 32? Can we get the actual connection pool size somehow?

gdcolella 21 minutes ago  Author
We can't get the actual one that's used (if it's set by dial option). The default is 4, so maybe 8 requests per connection is overkill. We could lower it to 16? I'd like to have more than 1 per connection to get better odds of hitting each one. Doing this correctly requires changes lower down, which I did implement somewhere else but is a lot more complicated.